repository-s3/src/main/java/io/aiven/elasticsearch/repositories/s3/S3RepositoryStorageIOProvider.java

```diff
@@ -0,0 +1,270 @@
+                @Override
+                public List<String> get(final int index) {
+                    final var start = index * MAX_DELETE_BULK;
+                    final var end = Math.min(start + MAX_DELETE_BULK, blobNames.size());
+                    return blobNames.subList(start, end);
+                }
+
+                @Override
+                public int size() {
+                    return (int) Math.ceil((double) blobNames.size() / (double) MAX_DELETE_BULK);
+                }
+            };
+
+            for (var idx = 0; idx < partitionList.size(); idx++) {
+                final var chunk = partitionList.get(idx);
+                final var deleteObjectsRequest =
+                        new DeleteObjectsRequest(bucketName)
+                                .withKeys(chunk.toArray(new String[0]));
+                try {
+                    Permissions.doPrivileged(() -> client.deleteObjects(deleteObjectsRequest));
+                } catch (final MultiObjectDeleteException e) {
+                    final var deleteErrorMessagePattern = "Couldn't delete object: %s. Reason: [%s] %s";
+                    e.getErrors()
+                            .stream()
+                            .map(de ->
+                                    String.format(
+                                            deleteErrorMessagePattern,
+                                            de.getKey(),
+                                            de.getCode(),
+                                            de.getMessage())
+                            )
+                            .forEach(logger::warn);
+                } catch (final AmazonClientException e) {
+                    throw new IOException("Couldn't delete objects: " + blobNames, e);
+                }
+            }
+        }
+
+        @Override
+        public List<String> listDirectories(final String path) throws IOException {
+            try {
+                return Permissions.doPrivileged(() -> list(listOfObjectsRequest(path)))
+                        .stream()
+                        .flatMap(listing -> listing.getCommonPrefixes().stream())
+                        .map(prefix -> prefix.substring(path.length()))
+                        .filter(Predicate.not(String::isEmpty))
+                        .map(prefix -> prefix.substring(0, prefix.length() - 1))
+                        .collect(Collectors.toList());
+            } catch (final AmazonClientException e) {
+                throw new IOException("Couldn't get list of directories for path " + path, e);
+            }
+        }
+
+        @Override
+        public Map<String, Long> listFiles(final String path, final String prefix) throws IOException {
+            try {
+                final var fullPath = path + prefix;
+                return Permissions.doPrivileged(() -> list(listOfObjectsRequest(fullPath)))
+                        .stream()
+                        .flatMap(l -> l.getObjectSummaries().stream())
```

Comment:
The same comment about `getObjectSummaries` result potentially being truncated.

Reply:
same here